string/string.cc

```diff
@@ -1,12 +1,113 @@
 #include "string.h"
 
-String::String()
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <string>
+
+
+const char* EMPTY = "";
+
+
+size_t strlen(const char* s)
+{
+	size_t len = 0;
+
+	while (s[len] != '\0')
+		len++;
+
+	return len;
+}
+
+
+String::String() {}
+
+String::~String() {}
+
+String::String(const String& o)
+{
+	mSize = o.size();
+
+	mValue = char_ptr(new char[mSize + 1]);
+
+	std::copy_n(o.begin(), mSize, begin());
+
+	mValue[mSize + 1] = '\0';
+}
+
+String::String(const char* s)
+{
+	mSize = strlen(s);
+
+	mValue = char_ptr(new char[mSize + 1]);
+
+	std::copy_n(s, mSize + 1, begin());
+}
+
+template<typename It> String::String(It first, It last)
 {
-	// TODO your code goes here, if any
+	mSize = std::distance(first, last);
+
+	mValue = char_ptr(new char[mSize + 1]);
+
+	std::copy_n(first, mSize, begin());
```

Comment:
Here you can't assume that the terminating zero is here.
Please note: this constructor, ans the other one share almost all of the code. Maybe you could implement the other ones in terms of this?
(Google: delegating constructor)